Added publish subscribe

wajd opened the pull request:
I added our project that adds publish subscribe functionality as a new module. This is still unfinished and would require further work but most core functionality is available. 
You can test it with this [Coap pubsub broker](https://github.com/PethrusG/CoAP-PubSub-Broker). 
 
refer to this [issue](https://github.com/eclipse/californium/issues/956) for further info or to the readme .

boaks commented on cf-pubsub/src/main/java/org/eclipse/californium/pubsub/Converter.java:
```diff
@@ -0,0 +1,264 @@
+/*******************************************************************************
+ * Copyright (c) 2018 RISE SICS and others.
```
I'm not sure, if your work for the RISE SICS as well. 
If not, just place your university or employer.

wajd replied:
We've asked the Network Systems Lab and are awaiting their response on what to put under the copyright. I apologize for forgetting to change it.

boaks replied:
Any news on that?
If the "Network Systems Lab" doesn't claim the "copyright",  you may also just add the names of the authors. But for a clean legal state, It would be also great, if such a "no-claim" is written in some e-mail :-).

wajd replied:
The networks lab and KTH (same entity) both said they have no copyright. As for us, the authors, we've already signed our names. So I think it should be fine if the RISE copyright is removed and just our names left as authors.

boaks replied:
The eclipse foundation requires a explicit copyright statement.
It's not big thing, just list the same names above instead of the "RISE SICS".

boaks commented:
Do you use the eclipse IDE? If so, would it be possible to import the source formatter "eclipse-formatter-profile.xml" and apply it to your java files?

boaks commented on cf-pubsub/pom.xml:
```diff
@@ -0,0 +1,67 @@
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-api</artifactId>
+        </dependency>
+        <dependency>
```
hamcrest and junit are currently not used.

slf4j-api also not.

Consider to remove these dependencies, if they are not used.

boaks commented on cf-pubsub/pom.xml:
```diff
@@ -0,0 +1,67 @@
+        <dependency>
+            <groupId>ca.jimr</groupId>
+            <artifactId>gae-mini-profiler</artifactId>
+            <version>2.0.3</version>
```
At least to compile, this seems to be not used,
Consider to change the scope of the dependency to runtime by adding `<scope>runtime</scope>`

boaks commented on cf-pubsub/src/main/java/org/eclipse/californium/pubsub/Converter.java:
```diff
@@ -0,0 +1,264 @@
+     * @return an array of WebLinks
+     */
+    public static WebLink[] getArray(Set<WebLink> webLinks) {
+        return webLinks.toArray(new WebLink[0]);
```
Seems to be equivalent to `return new WebLink[0];`

wajd replied:
This function might actually not be needed if this class is to be added to Californium, as someone using Californium should know how to do toArray properly on a Set of WebLink.

boaks commented on cf-pubsub/src/main/java/org/eclipse/californium/pubsub/Converter.java:
```diff
@@ -0,0 +1,264 @@
+     * @param webLink WebLink
+     * @return an int ContentType
+     */
+    public static int getCT(WebLink webLink) {
```
Usually we try not to use abbreviations. Please consider to rename it to `getContentType`.

boaks replied:
Also for `setCT` an `getCTString`.

boaks commented on cf-pubsub/src/main/java/org/eclipse/californium/pubsub/PubSub.java:
```diff
@@ -0,0 +1,343 @@
+            req.setURI(client.getURI());
+            req.setObserve();
+
+            config.set(NetworkConfig.Keys.TOKEN_SIZE_LIMIT, 4);
```
I would recommend to either move this into the constructor, or use a properties file and 
https://github.com/eclipse/californium/blob/2.0.x/demo-apps/cf-extplugtest-server/src/main/java/org/eclipse/californium/extplugtests/ExtendedTestServer.java#L68 with
https://github.com/eclipse/californium/blob/2.0.x/demo-apps/cf-extplugtest-server/src/main/java/org/eclipse/californium/extplugtests/ExtendedTestServer.java#L114 .

boaks commented on cf-pubsub/src/main/java/org/eclipse/californium/pubsub/PubSub.java:
```diff
@@ -0,0 +1,343 @@
+            config.set(NetworkConfig.Keys.TOKEN_SIZE_LIMIT, 4);
+            RandomTokenGenerator rand = new RandomTokenGenerator(config);
+            Token token = rand.createToken(false);
+            req.setToken(token);
```
If you set TOKEN_SIZE_LIMIT in the "Standard" NetworkConfig (as done in line 321), it should be effective without generating a token on your own.

boaks commented on cf-pubsub/src/main/java/org/eclipse/californium/pubsub/PubSub.java:
```diff
@@ -0,0 +1,343 @@
+        public void unsubscribe() {
+            if (this.relation != null) {
+                relation.proactiveCancel();
+                int mid = relation.getCurrent().advanced().getMID();
```
That creates a very rare race condition. 
Line 335 may have already changed the current response and no further response will be received.
The while may then last for longer ...

boaks commented on cf-pubsub/src/main/java/org/eclipse/californium/pubsub/Topic.java:
```diff
@@ -0,0 +1,202 @@
+    public Topic(WebLink wl) {
+        this.uri = wl.getURI().substring(1).split("(/)");
+        this.name = this.uri[this.uri.length - 1];
+        String corchete = wl.toString().substring(wl.toString().indexOf('[') + 1);
```
I'm not sure, if this is reliable. It seems to read the first attribute value list, regardless of the attributes name. I would prefer a solution, which gets the attribute-value by name.

If that string-search approach is intended, then I would prefer to call `wl.toString()` only once and re-use it. 

Or

boaks commented on cf-pubsub/src/main/java/org/eclipse/californium/pubsub/Topic.java:
```diff
@@ -0,0 +1,202 @@
+        try {
+            this.ct = Integer.parseInt(corchete.substring(0, 1));
+        } catch (NullPointerException | NumberFormatException | ArrayIndexOutOfBoundsException e) {
+            this.ct = Integer.parseInt(corchete.substring(0, 0));
```
Parse a empty string? 

> Exception in thread "main" java.lang.NumberFormatException: For input string: ""

boaks commented on cf-pubsub/src/main/java/org/eclipse/californium/pubsub/Topic.java:
```diff
@@ -0,0 +1,202 @@
+     * @param topic a Topic which you want to get the subtopics and parent Topics from
+     * @param at    an already existing ArrayList of Topics where you want all the Topics to be
+     */
+    public void getTopics(Topic topic, ArrayList<Topic> at) {
```
May be consider to make this function `static`?

boaks commented on cf-pubsub/src/main/java/org/eclipse/californium/pubsub/Topic.java:
```diff
@@ -0,0 +1,202 @@
+     * @return true if given Topic has same URI as instance Topic
+     */
+    public boolean equals(Topic topic) {
+        return (this.getName().equals(topic.getName()) && this.getURIString().equals(topic.getURIString()));
```
the name is the last part of the URI? So no need to compare it twice?

Try `Arrays.equals(this.uri, topic.uri);`

boaks commented on cf-pubsub/src/main/java/org/eclipse/californium/pubsub/Topic.java:
```diff
@@ -0,0 +1,202 @@
+
+/**
+ * The TopicComparator class is used for sorting purposes whenever it is needed
+ * to compare and sort Topics. It sorts in terms of alphabetical order and URI length
```
It's ok, not have unit test, but here I'm sure, they would show a defect.

boaks commented:
I'm not sure, how much time you can spend, but the

> Copyright (c) 2018 RISE SICS and others.

must use either your university or your employer.

wajd commented:
> Do you use the eclipse IDE? If so, would it be possible to import the source formatter "eclipse-formatter-profile.xml" and apply it to your java files?

We used IntelliJ unfortunately.

wajd commented:
> I'm not sure, how much time you can spend to go through the list above. If you can't fix all, that will be also OK. But the
> 
> > Copyright (c) 2018 RISE SICS and others.
> 
> must use either your university or your employer.

I apologize for burdening you further but it seems I have no time to go through the requested changes. If you can make those changes yourself, please go ahead. I've already replied above about the copyright claim.

boaks commented:
> I apologize for burdening you further but it seems I have no time to go through the requested changes. If you can make those changes yourself, please go ahead. I've already replied above about the copyright claim.

Except the copyright, that will be OK.
Please just copy the list of authors to the copyright above. Then squash your 2 commits into 1 and force push that to your PR branch. After that I'm able to create the CQ.

wajd commented:
I updated the copyright, removed the unneeded dependencies, and renamed the CT methods. I did a squash and I hope I did it right since this is the first time I've done a git squash.

boaks commented:
Squash is "well done" :-) .

But the copyright should end with "and others."
Would it be possible for you to add that? 
Amend that changes to your last commit and force push it again.

wajd commented:
done.

boaks commented:
Thanks!

I will create a CQ. Please don't update this branch any more.
Changes and fixes must be in separate PRs.

Just to mention:
If the contribution is done by one person and below 1000 lines, we can "by pass" the CQ. 
Therefore, if you or one of your team want to contribute some improvements, it's easier, if that is done "alone" :-).

boaks commented:
Hi all, thanks a lot for your contribution and patience!
